### PR TITLE
[CORDA-3082] Update app upgrade notes to document source incompatibility

### PR DESCRIPTION
**Pull request by JamesHR3**
The introduction of the `Destination` interface has introduced a potential source incompatibility between platform versions 4 and 5. This occurs due to a type inference limitation in the Kotlin compiler. This PR updates the upgrade notes for platform version 5 to describe the possible break and how to fix it.

### CODE REVIEW

File: docs/source/app-upgrade-notes.rst [hunk 1]
```diff
@@ -23,6 +23,73 @@ from the new features in the latest release.
...
+        Pair(lender, ourIdentity)
+    }
+
+    val obligation = Obligation(100.dollars, lenderId, borrowerId)
+
+Compiling this code against Platform Version 5 will result in the following error:
+
+`Type mismatch: inferred type is Any but AbstractParty was expected`
+
+The issue here is that a new interface introduced in Platform Version 5 can cause type inference failures when a variable is used as an
```
Comment (gendal) [inline docs/source/app-upgrade-notes.rst]: link to the section of the release notes - or kdocs/javadocs?

Comment (JamesHR3) [reply docs/source/app-upgrade-notes.rst]: I've linked to the changelog where this is currently mentioned. I can also link to the javadocs if that would be helpful?

File: docs/source/app-upgrade-notes.rst [hunk 2]
```diff
@@ -23,6 +23,74 @@ from the new features in the latest release.
...
+
+.. note:: If you are upgrading from a platform version older than 4, then the upgrade notes for upgrading to Corda 4 (below) also apply.
+
+Step 1. Handle any source compatibility breaks
+----------------------------------------------
+
+The following code, which compiled in Platform Version 4, will not compile in Platform Version 5:
+
+.. code-block:: kotlin
+
```
Comment (lankydan) [inline docs/source/app-upgrade-notes.rst]: Is this only an issue in Kotlin? I think it probably is, so should we mention that so java devs don't spend time checking if they need to change anything. If it also affects java code, then I think we should include java samples as well.

File: docs/source/app-upgrade-notes.rst [hunk 3]
```diff
@@ -23,6 +23,74 @@ from the new features in the latest release.
...
+        Pair(anonymousIdentitiesResult[lenderSession.counterparty]!!, anonymousIdentitiesResult[ourIdentity]!!)
+    } else {
+        Pair(lender, ourIdentity)
+    }
+
+    val obligation = Obligation(100.dollars, lenderId, borrowerId)
+
+Compiling this code against Platform Version 5 will result in the following error:
+
+`Type mismatch: inferred type is Any but AbstractParty was expected`
```
Comment (lankydan) [inline docs/source/app-upgrade-notes.rst]: I think we should be using double backticks for code to use proper syntax highlighting (currently they are rendered as italics).

File: docs/source/app-upgrade-notes.rst [hunk 4]
```diff
@@ -23,6 +23,74 @@ from the new features in the latest release.
...
+To fix this, an explicit type hint must be provided to the compiler:
+
+.. code-block:: kotlin
+
+    data class Obligation(val amount: Amount<Currency>, val lender: AbstractParty, val borrower: AbstractParty)
+
+    val (lenderId, borrowerId) = if (anonymous) {
+        val anonymousIdentitiesResult = subFlow(SwapIdentitiesFlow(lenderSession))
+        Pair(anonymousIdentitiesResult[lenderSession.counterparty]!!, anonymousIdentitiesResult[ourIdentity]!!)
+    } else {
```
Comment (lankydan) [inline docs/source/app-upgrade-notes.rst]: maybe a single comment in the code highlighting the `Pair` which now has a type will make the example clearer.

Something like:

```kotlin
// The Pair now defines it's generic types
Pair<AbstractParty, AbstractParty>(lender, ourIdentity)
```

File: docs/source/app-upgrade-notes.rst (hunk 2 above)
Comment (JamesHR3) [reply docs/source/app-upgrade-notes.rst]: I think it is Kotlin specific, my understanding is that the Java compiler can work these situations out. I've called this out in the section title and in the main text.

**Comment by rick-r3** [comment]
I'm not sure that explanation is quite right, is it ?  I thought the explanation was that the kotlin _language_ doesn't (currently) support an inferred type with two ancestors and so resolves to the first single ancestor (Any).  The compiler doesn't get a choice.  Java does support multiple ancestors, and thus has no problems.

We need to be very precise in our wording.  The kotlin devs can read this PR and the eventual commit.

@gendal @GavT Are we worried about any wording here from a legal standpoint ?  Currently reads like a criticism of the kotlin compiler to me.

**Comment by gendal** [comment]
I'm not worried from a legal perspective but we should obviously endeavour to be technically correct... if the wording is wrong let's fix it.

**Comment by mikehearn** [comment]
Have we asked JetBrains if this is the real cause?

**Comment by shamsasari** [comment]
Is this PR waiting on someone?

**Comment by mikehearn** [comment]
Not sure. I didn't mean to block it.

**Comment by JamesHR3** [comment]
Suggestion is that this might be fixed in 1.3: https://discuss.kotlinlang.org/t/type-inference-problem-with-classes-with-base-class-and-interface/13894

**Pull request merged**